p2pclient/unary_handlers.go

```diff
@@ -0,0 +1,279 @@
+package p2pclient
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/google/uuid"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	pb "github.com/libp2p/go-libp2p-daemon/pb"
+
+	ggio "github.com/gogo/protobuf/io"
+	"github.com/gogo/protobuf/proto"
+)
+
+type persistentConnectionResponseFuture chan *pb.PersistentConnectionResponse
+
+type UnaryHandlerFunc func(context.Context, []byte) ([]byte, error)
+
+func (u UnaryHandlerFunc) handle(ctx context.Context, w ggio.Writer, req *pb.PersistentConnectionResponse) {
+	result, err := u(ctx, req.GetRequestHandling().Data)
+
+	response := &pb.CallUnaryResponse{}
+	if err == nil {
+		response.Result = &pb.CallUnaryResponse_Response{
+			Response: result,
+		}
+	} else {
+		response.Result = &pb.CallUnaryResponse_Error{
+			Error: []byte(err.Error()),
+		}
+	}
+
+	w.WriteMsg(
+		&pb.PersistentConnectionRequest{
+			CallId: req.CallId,
+			Message: &pb.PersistentConnectionRequest_UnaryResponse{
+				UnaryResponse: response,
+			},
+		},
+	)
+}
+
+func (c *Client) run(r ggio.Reader, w ggio.Writer) {
+	for {
+		var resp pb.PersistentConnectionResponse
+		r.ReadMsg(&resp)
+
+		callID, err := uuid.FromBytes(resp.CallId)
+		if err != nil {
+			log.Debugw("received response with bad call id:", "error", err)
+			continue
+		}
+
+		switch resp.Message.(type) {
+		case *pb.PersistentConnectionResponse_RequestHandling:
```

Comment:
It is misleading to have a response type that actually corresponds to a request :)

I'd suggest to combine `PersistentConnectionRequest` and `PersistentConnectionResponse` to a single `PersistentConnectionMessage` type (they are already very similar) and send protobufs of this type both ways.

Reply:
I like the suggestion, but I would rather do it in a separate PR, as this would require also updating some code in already approved PR's in the python client.

Reply:
Okay, I will add a todo item to the description :)